torch/csrc/jit/passes/memory_planning.cpp

```diff
@@ -0,0 +1,399 @@
+bool overlapLiveRange(
+    const UniqueLiveRange& ulvr1,
+    const UniqueLiveRange& ulvr2) {
+  return overlap(
+             ulvr1.lvr.begin, ulvr1.lvr.end, ulvr2.lvr.begin, ulvr2.lvr.end) <=
+      0;
+}
+
+// since memory address are zero indexed, offset + size ends at the *beginning*
+// of the (offset+size)th byte. hence overlap is like open intervals (i.e.
+// overlap is more than at endpoints)
+bool overlapMemRegion(const MemRegion& reg1, const MemRegion& reg2) {
+  return overlap(
+             reg1.offset,
+             reg1.offset + reg1.size,
+             reg2.offset,
+             reg2.offset + reg2.size) < 0;
+}
+
+bool overlapAllocs(const MemAllocation& m1, const MemAllocation& m2) {
+  return overlapLiveRange(m1.ulvr, m2.ulvr) && overlapMemRegion(m1.reg, m2.reg);
+}
+
+// stack all tensors end to end "as you see them" over the entire lifespan of
+// the plan
+std::vector<MemAllocation> naive(
+    SortedLiveRangeMap<size_t> managed_live_ranges) {
+  std::vector<MemAllocation> allocations;
+  allocations.reserve(managed_live_ranges.size());
+  size_t offset = 0;
+  for (const auto& item : managed_live_ranges) {
+    auto ulvr = item.first;
+    auto size = item.second;
+    auto id = item.first.id;
+    auto aligned_size = MemoryPlanner::computeAlignedTensorSize(size);
+    allocations.push_back({ulvr, {offset, aligned_size}});
+    offset += aligned_size;
+  }
+  return allocations;
+}
+
+c10::optional<size_t> computeStorageSize(const Value& value) {
+  auto node_header = getHeader(value.node());
+  auto ttp = value.type()->cast<TensorType>();
+  if (!ttp) {
+    TORCH_WARN(
+        node_header,
+        "; ",
+        value.debugName(),
+        " isn't a tensortype: ",
+        *value.type());
+    return c10::nullopt;
+  }
+  if (!ttp->scalarType().has_value()) {
+    TORCH_WARN(
+        node_header,
+        "; ",
+        value.debugName(),
+        " was profiled but didn't have a scalar type: ",
+        *ttp);
+    return c10::nullopt;
+  }
+  if (!ttp->sizes().concrete_sizes().has_value()) {
+    TORCH_WARN(
+        node_header,
+        "; ",
+        value.debugName(),
+        " was profiled but doesn't have sizes: ",
+        *ttp);
+    return c10::nullopt;
+  }
+
+  auto scalar_type = ttp->scalarType();
+  if (!scalar_type.has_value()) {
+    TORCH_WARN(
+        node_header,
+        "; ",
+        value.debugName(),
+        " doesn't have a scalar type: ",
+        *ttp);
+    return c10::nullopt;
+  }
+
+  auto element_size = c10::elementSize(scalar_type.value());
+  // TODO: when does this fail? answer: in place mutation
+  auto numel = ttp->numel();
+  if (!numel.has_value()) {
+    TORCH_WARN(
+        node_header, "; ", value.debugName(), " doesn't have numel: ", *ttp);
+    return c10::nullopt;
+  }
+
+  return numel.value() * element_size;
+}
+
+Node* insertSlabNode(
+    const std::shared_ptr<Graph>& graph,
+    size_t total_size,
+    c10::optional<at::Device> device_type = c10::nullopt) {
```

Comment:
Maybe this should be required input ?

Reply:
not sure. we do have that "pickDevice" method from tensorexprs (@Krovatkin was the one that recommended using it). but at the same time explicitness is good